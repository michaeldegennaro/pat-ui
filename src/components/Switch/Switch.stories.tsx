import React from 'react';
import { action } from '@storybook/addon-actions';
import Switch from './Switch';

export default {
  title: 'Switch',
  component: Switch,
};

const buttonStyle: React.CSSProperties = {
  marginRight: '5px',
  marginTop: '5px',
};

export const DefaultSwitch = () => (
  <Switch onClick={action('Default Switch clicked')}
  >Default Switch</Switch>
);

export const DisabledSwitch = () => (
    <div>
    <Switch onClick={action('Activated Switch clicked')}
>Enabled</Switch>
    <Switch disabled={true}>Disabled</Switch>
    </div>
  );

export const DifferentSizeSwitch = () => (
  <div>
    <Switch
      switchSize='small'
      onClick={action('Small Switch clicked')}
    >
      Small Switch
    </Switch>
    <Switch
      switchSize='medium'
      onClick={action('Medium Switch clicked')}
    >
      Default Switch
    </Switch>

    <Switch
      switchSize='large'
      onClick={action('Large Switch clicked')}
    >
      Large Switch
    </Switch>
  </div>
);

export const DiffTypeSwitch = () => (
  <div>
    <Switch
      onClick={action('Default Switch clicked')}
    >
      Primary Switch
    </Switch>
    <Switch
      textPosition='top'
      onClick={action('Secondary Switch clicked')}
    >
      Secondary Switch
    </Switch>
  </div>
);