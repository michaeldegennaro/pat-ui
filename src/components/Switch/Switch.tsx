//import "./_Switch.css";
import React, { FC} from "react";
import { classNames } from '../../utils/classNames';


export type Size = 'large' | 'small' | 'medium';
export type TextPosition = 'default' | 'top' | 'left' ;


export interface ISwitchProps {
  /** set classname */
  className?: string;
  /** set switch size */
  switchSize?: Size;
  /** set text positioning */
  textPosition?: TextPosition;
  /** set action on switch clicked */
  onClick?: () => void;
  /** set if checked */
  checked?: boolean;
  /** set if disabled */
  disabled?: boolean;
}


export type patSwitchProps = ISwitchProps;

/**
 * A Switch indicates a possible user action.
 *
 * ```js
 * import {Switch} from 'pat-ui'
 * ```
 */
export const Switch: FC<patSwitchProps> = (props) => {
  const { switchSize, textPosition, children, disabled, className, checked, ...rest } = props;

  let styleClasses = classNames('card', {
    [`${switchSize}`]: true,
    [`${textPosition}`]: true,
  });

  if (className) {
    styleClasses += ' ' + className;
}

  return (
    <div className={"container " + textPosition} data-testid='switch-element'>
      <label className={'switch ' + styleClasses + ' ' + props.disabled} data-testid='switch-label'>
        <input type="checkbox" onClick={props.onClick} />
        <span className={'slider switch ' + switchSize + ' ' + props.disabled}></span>
      </label>
      <span className={'label ' + props.disabled + ' ' + styleClasses}>{props.children}</span>
    </div>
  );
}
Switch.defaultProps = {
  switchSize: 'medium',
  textPosition: 'default'
};
export default Switch;

