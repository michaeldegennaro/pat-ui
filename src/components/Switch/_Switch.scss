.container {
  display: flex;
  justify-content: center;
  margin-bottom: 5px;
  font-family: sans-serif;
  text-align: center;
  &.top {
    flex-direction: column-reverse;
    margin-bottom: 17px;
  }
  &.left {
    flex-direction: row-reverse;
  }
}

.switch {
  position: relative;
  display: inline-block;
  margin-bottom: 20px;
  width: 45px;
  height: 15px;
  top: 3px;
  input {
    opacity: 0;
    width: 0;
    height: 0;
  }
  &.top{
    margin: auto;
  }
  &.small{
    @include switch-size(
      $switch-width-small,
      $switch-height-small,
      $switch-pos-small
        );
  }
  &.large {
    @include switch-size(
  $switch-width-large,
  $switch-height-large,
  $switch-pos-large
    );
  }
}

.label {
  position: relative;
  bottom: 4px;
  left: 10px;
  font-size: 20px;
  &.true {
    color: $gray-500;
  }
  &.top{
    left:0px;
  }
  &.left{
    left:-10px;
  }
  &.small{
    bottom: 4.5px;
  }
  &.large{
    bottom: 2.5px;
  }
}

.slider {
  position: absolute;
  cursor: pointer;
  top: 0;
  left: 0;
  right: 0;
  bottom: 0;
  background-color: $gray-400;
  -webkit-transition: 0.4s;
  transition: 0.4s;
  &.true {
    background-color: $slider-bakcground-color;
    cursor: not-allowed;
  }
  &:before{
    position: absolute;
    content: "";
    height: 20px;
    width: 20px;
    left: 4px;
    bottom: -3px;
    background-color: $blue;
    -webkit-transition: 0.4s;
    transition: 0.5s;
  }
  &.small:before{
    @include switch-slider(
      $slider-width-small,
      $slider-height-small,
      $slider-left-small,
      $slider-bottom-small,
    );
  }
  &.large:before{
    @include switch-slider(
      $slider-width-large,
      $slider-height-large,
      $slider-left-large,
      $slider-bottom-large,
    );
  }
  &.true:before {
    background-color: $slider-checked-color;
  }
}

input{
  &:focus + .slider{
    box-shadow: 0 0 3px $blue;
  }
  &:checked + .slider{
    background-color: $slider-checked-color;
  }
  &:checked + .slider.true{
    background-color: $slider-bakcground-color;
  }
  &:focus + .slider.true {
    box-shadow: 0 0 0px $slider-bakcground-color;
  }
  &:checked + .slider:before {
    transform: translateX(17px);
  }
  &:checked + .slider.small:before {
    transform: translateX(11px);
  }
  &:checked + .slider.large:before {
    transform: translateX(19px);
  }
  &:checked + .slider.true:before {
    transform: translateX(0px);
  }
}

.slider {
  border-radius: 34px;
  &:before{
    border-radius: 50%;
  }
}
