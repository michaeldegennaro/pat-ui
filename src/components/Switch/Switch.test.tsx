import React from 'react';
import {render, fireEvent, screen} from '@testing-library/react';
import Switch ,{ ISwitchProps } from './Switch';

describe('Switch', () => {
    it('should match snapshot', () => {
        const { asFragment } = render(<Switch> Snapshot Switch </Switch>);
        expect(asFragment()).toMatchSnapshot();
      });

    it('should render default switch', () => {
        // Default message
        const switchBasicProps: ISwitchProps = {
            switchSize: 'medium',
            textPosition: 'default',
            className: 'test'
        }
        const msgBasicWrapper = render(<Switch {...switchBasicProps}></Switch>);
        const msgElement = screen.getByTestId('switch-element');
        expect(msgElement).toHaveClass('container default');
    });
    it('should render default switch', () => {
        // Default message
        const switchLargeProps: ISwitchProps = {
            switchSize: 'large',
            textPosition: 'top',
            disabled: true,
            onClick: () => {
                console.log('Here')
            },
            className: 'test-two'
        }
        const msgBasicWrapper = render(<Switch {...switchLargeProps}></Switch>);
        const msgElement = screen.getByTestId('switch-label');
        expect(msgElement).toHaveClass('switch large top test-two true');
    });


});
