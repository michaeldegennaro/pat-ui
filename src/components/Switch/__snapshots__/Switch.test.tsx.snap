// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Switch should match snapshot 1`] = `
<DocumentFragment>
  <div
    class="container default"
    data-testid="switch-element"
  >
    <label
      class="switch card medium default undefined"
      data-testid="switch-label"
    >
      <input
        type="checkbox"
      />
      <span
        class="slider switch medium undefined"
      />
    </label>
    <span
      class="label undefined card medium default"
    >
       Snapshot Switch 
    </span>
  </div>
</DocumentFragment>
`;
